File: 2025-season6/3주차/최영권/ReverseLinkedList.java
package LinkedList;

public class ReverseLinkedList {
    public static void main(String[] args) {
        ReverseLinkedList reverseLinkedList = new ReverseLinkedList();
        ListNode listNode = reverseLinkedList.reverseList(new ListNode(1, new ListNode(2, new ListNode(3, new ListNode(4, new ListNode(5))))));
        System.out.println(listNode);
    }

    // 시간복잡도 : 연결리스트 길이가 n 일때, O(n)
    public ListNode reverseList(ListNode head) {
        ListNode prev = null;
        ListNode curr = head;
        while (curr != null) {
            ListNode next = curr.next;
            curr.next = prev;
            prev = curr;
            curr = next;
        }
        return prev;
    }

    public static class ListNode {
        @Override
        public String toString() {
            return "ListNode{" +
                    "val=" + val +
                    ", next=" + next +
                    '}';
        }

        int val;
        ListNode next;

        ListNode() {
        }

        ListNode(int val) {
            this.val = val;
        }

        ListNode(int val, ListNode next) {
            this.val = val;
            this.next = next;
        }


    }
}

File: 2025-season6/3주차/이영범/reverse-linked-list.ts

/*
206. Reverse Linked List
Easy

Given the head of a singly linked list, reverse the list, and return the reversed list.
 

Example 1:


Input: head = [1,2,3,4,5]
Output: [5,4,3,2,1]
Example 2:


Input: head = [1,2]
Output: [2,1]
Example 3:

Input: head = []
Output: []
 

Constraints:

The number of nodes in the list is the range [0, 5000].
-5000 <= Node.val <= 5000
 

Follow up: A linked list can be reversed either iteratively or recursively. Could you implement both?
*/
// Definition for singly-linked list.
export class ListNode {
    val: number
    next: ListNode | null
    constructor(val?: number, next?: ListNode | null) {
        this.val = (val===undefined ? 0 : val)
        this.next = (next===undefined ? null : next)
    }
}

function reverseList(head: ListNode | null): ListNode | null {
    return reverseListRecursively(head)
};

/*
    최종 TC: O(n)
    최종 SC: O(1)
*/
function reverseListIteratively(head: ListNode | null): ListNode | null {
    if (!head || !head.next) return head

    let prev: ListNode | null = null
    // TC: O(n)
    while (head) {
        const next = head.next // 다음 백업
        head.next = prev // 링크 뒤집기
        prev = head // prev 갱신
        head = next // 백업한 다음으로 갱신
    }
    return prev
}

/*
    최종 TC: O(n)
    최종 SC: O(n)
*/
function reverseListRecursively(head: ListNode | null): ListNode | null {
    if (!head || !head.next) return head
    let prev: ListNode | null = null
    // TC: O(n)
    // SC: O(n)
    return reverse(head, prev)
}

function reverse(head: ListNode | null, prev: ListNode | null) {
    if (!head) return prev
    
    const next = head.next // 다음 백업
    head.next = prev // 링크 뒤집기
    prev = head // prev 갱신
    head = next // 백업한 다음으로 갱신
    
    return reverse(head, prev)
}